### ISSUE TITLE
Filtering images. Something needs too be done. 

### ISSUE THREAD

**Issue by tonyram57** [FrostCo/AdvancedProfanityFilter#532]
Really your extention is working great and I can read articles and forums great without profanity but I know we talked about this before but the images are really annoying me. I been encountered much profanity in images basically the f word everytime the last four days so I am just saying instead of trying to focus on trying to using the mute audio feature that never once ever worked you really need to staer to try to get it work on images even if it's a beta feature that needs work. Thanks and happy new year.

**Comment by tonyram57** [comment]
The f word in images is really ticking me off and I am encountering nearly everyday 90% of the time in images.

**Comment by richardfrost** [comment]
Sorry @tonyram57, unfortunately I haven't had much time to look into this yet. Its on my list to get to it, but I can't promise when that will be. If you could share a couple of these images with me it would help me see how well it might work.

Using [tesseract](https://tesseract.projectnaptha.com/) you can see what the results might be using their library, which is probably the best one freely available that could be included in the extension. Here are some examples:

## Best (still not perfect)
![Good - Screenshot 2024-01-04 105327](https://github.com/FrostCo/AdvancedProfanityFilter/assets/1507205/c9bf3cdd-b408-42ec-aafe-178e5fbdf80c)
![good - Screenshot 2024-01-04 105145](https://github.com/FrostCo/AdvancedProfanityFilter/assets/1507205/3ee008e1-7cd1-4252-acbd-4525aa8f138b)
gg

## Partially worked
![Ok - Screenshot 2024-01-04 105723](https://github.com/FrostCo/AdvancedProfanityFilter/assets/1507205/b2e01097-c146-4388-9337-3553fd47345d)
![ok - Screenshot 2024-01-04 105558](https://github.com/FrostCo/AdvancedProfanityFilter/assets/1507205/c715347c-f758-47cf-a9b3-51adcae94ac2)
![Ok - Screenshot 2024-01-04 105429](https://github.com/FrostCo/AdvancedProfanityFilter/assets/1507205/85336b01-6575-4f0d-a9fe-1d9f179b8892)
![ok - Screenshot 2024-01-04 104936](https://github.com/FrostCo/AdvancedProfanityFilter/assets/1507205/f40eeff0-e7c2-46a0-b0a1-52b4581b3bd6)

## Total failure
![bad - Screenshot 2024-01-04 104309](https://github.com/FrostCo/AdvancedProfanityFilter/assets/1507205/cbc574fa-4a8d-4c19-83bc-8816f797a0db)
![Screenshot 2024-01-04 105526](https://github.com/FrostCo/AdvancedProfanityFilter/assets/1507205/fae68841-e279-4fe6-a493-71e09126c94c)
![Screenshot 2024-01-04 105658](https://github.com/FrostCo/AdvancedProfanityFilter/assets/1507205/f58cdae1-d198-4246-a1d3-a0fb569097b1)
![Screenshot 2024-01-04 110152](https://github.com/FrostCo/AdvancedProfanityFilter/assets/1507205/c82d60d4-44d6-4cd9-94ea-574c7001560f)
![Screenshot 2024-01-04 110217](https://github.com/FrostCo/AdvancedProfanityFilter/assets/1507205/327596d6-6cf9-4bcc-8cd2-045b46960fb6)

I'm curious to know how some of your examples fare, but based on these results I'm not holding my breath it will be super reliable. You can either test them yourself on that website, or share the images with me and I can give it a try.

**Comment by tonyram57** [comment]
Thanks for the examples but those images almost always never contain profanity. You have not been on social media which is ironic because I recently meet some really good people who got me through stuff. Anyway, the  images three all have a similar structure that looks easier to detect. I will send some clean examples later.

**Comment by tonyram57** [comment]
Here is a type of image which is used often.
![example](https://github.com/FrostCo/AdvancedProfanityFilter/assets/39742654/4c502891-745b-47f1-8838-6598b3e04757)

**Comment by tonyram57** [comment]
![example2](https://github.com/FrostCo/AdvancedProfanityFilter/assets/39742654/a5d59fa1-e916-4254-961e-84c5de6f7a30)

**Comment by tonyram57** [comment]
![example3](https://github.com/FrostCo/AdvancedProfanityFilter/assets/39742654/d7791ec4-10e0-4e6f-99d2-a77376dd020b)

I will send more as I find them.

**Comment by tonyram57** [comment]
![Screenshot_2024-01-27-19-44-44-88_50ef9f5a0f3fc24b6f0ffc8843167fe4](https://github.com/FrostCo/AdvancedProfanityFilter/assets/39742654/acea1f73-ca4f-404f-aaff-cafe87fe65ec)

**Comment by richardfrost** [comment]
Thanks for sending these examples over @tonyram57. Here are the results running through the OCR process:

# Good
![Screenshot 2024-01-27 200324](https://github.com/FrostCo/AdvancedProfanityFilter/assets/1507205/9f4ddd0f-24f2-4e6a-9770-67b88d07820e)

# Pretty good
![Screenshot 2024-01-27 200400](https://github.com/FrostCo/AdvancedProfanityFilter/assets/1507205/8453a794-f113-4a45-8304-4da00bb31746)

# Okay
![Screenshot 2024-01-27 200102](https://github.com/FrostCo/AdvancedProfanityFilter/assets/1507205/4e8f967d-6cb2-4744-8d71-6e8db740999f)

# Failed
![Screenshot 2024-01-27 195938](https://github.com/FrostCo/AdvancedProfanityFilter/assets/1507205/068e43ef-08f0-4fcd-a9ec-9e7b811cb578)

As you can see, it is detecting quite a few wrong characters/words, which likely wouldn't cause too big of an issue for false positives when filtering because they usually aren't whole words.

The other part of this issue is that so far all this is focused on recognizing text. Once recognized, there won't really be a lot of options on what to do. It will not be possible to alter/filter the image. Options I can think of so far:
1. Block/hide the image
2. Blur the image and allow it to be shown by hovering/clicking/tapping on it
3. Replace the image with the filtered OCR text (May not be super helpful/useful based on the OCR results of these images having lots of extra garbage characters)

Would you still think it is worth it with one of those 3 options?

Another potential issue is speed/performance. These sample images have been taking between 0.5 - 3 seconds to process. This could really slow down performance on pages with heavy image use. I won't know how bad it is until I can play around with it in the actual extension.

**Comment by tonyram57** [comment]
Thanks for all your hard work. I did not realize how difficult filtering images would be. I thought it would be as simple as just detecting the text and then blocking the image. I forgot that the actual background image can trigger text which can make false positives.

It's up to you. I say you can go for it if you have the time but make it a beta feature you must turn on by default that will impact performance and may not work right. I am still going to visit social media where I encounter it the most. I made some real person friends there I connect too that hit me through hard times. I am not going to let the bad apples spoiled it for me. But anyway thanks for all your work.

**Comment by richardfrost** [comment]
@tonyram57 Yeah, it is a hard problem to solve. It reminds me of this funny comic: https://xkcd.com/1425/, which illustrates the issue pretty well. There are some things that seem hard that turn out to be very easy, and others that seem simple (like this one), but turn out to be almost impossible.

I'll keep it on the list to at least try with an integration in the code and see how it goes when I have some time.

I'm glad you found a good group of people though, even if you have to put up with seeing some things you don't want to see.

**Comment by tonyram57** [comment]
Thanks. The worst part about it is that the most profanity about 90% I see in images is f off. It seems that's very popular in images on social media which is annoying.

**Comment by richardfrost** [comment]
Yeah, that is annoying. Well, sorry I can't do anything with the filter yet. Maybe one day we'll be able to help with that issue.

**Comment by tonyram57** [comment]
Any progress? Images are quite annoying. However, enjoying movies and TV shows is a joy now that I figured out the caption option to only show when there is profanity **** but the other non swears, so I know what they say. I bought the Safari app to support the developer.

**Comment by richardfrost** [comment]
Thanks for your feedback and support @tonyram57, I appreciate it. There hasn't been much on this front unfortunately. I do hope one day to have something that will be able to accomplish this, but unfortunately I just don't have the time or resources right now to do much else with it. I'm keeping this issue open because it is definitely something I'd like to address in the future though.
